akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala

```diff
@@ -184,16 +184,23 @@ import akka.util.OptionVal
   }
 
   private def withSafelyAdapted[U, V](adapt: () => U)(body: U => V): Unit = {
-    try {
-      val a = adapt()
-      if (a != null) body(a)
-      else
-        ctx.log.warn(
-          "Adapter function returned null which is not valid as an actor message, ignoring. This can happen for example when using pipeToSelf and returning null from the adapt function. Null value is ignored and not passed on to actor.")
+    var failed = false
+    val adapted: U = try {
+      adapt()
     } catch {
       case NonFatal(ex) =>
         // pass it on through the signal handler chain giving supervision a chance to deal with it
         handleSignal(MessageAdaptionFailure(ex))
+        // Signal handler should actually throw so this is mostly to keep compiler happy (although a user could override
+        // the MessageAdaptionFailure handling to do something weird)
+        failed = true
+        null.asInstanceOf[U]
+    }
+    if (!failed) {
+      if (adapted != null) body(adapted)
+      else
+        ctx.log.warn(
+          "Adapter function returned null which is not valid as an actor message, ignoring. This can happen for example when using pipeToSelf and returning null from the adapt function. Null value is ignored and not passed on to actor.")
```

Comment:
This looks somewhat horrible, not entirely sure it is worth it, it avoids some wrapper allocation per message, but adapted messages anyways will lead to a bunch of allocations so perhaps a `Try[]` on top wouldn't really hurt.

The important part is to try-catch only the message adaptation and not the handling of the adapted message as that will already go through the behavior stack.

Reply:
We should avoid additional allocations for the normal path of message adapters, and I guess that is what the ugly part of this avoids. Looks ok to me.

Reply:
I don't see how this is solving the issue. Could you clarify?

In previous code `body(adapted)` was only called if `adapt()` was successful called.

That's the same here, except that we are using a boolean flag. 

How could that solve anything?

Reply:
In the previous any exception from both body() and adapt() was caught and handled the same way, now any failures from the body bubbles up to the default exception handling on message.

Reply:
Sorry, "bubbles up" is the wrong way to describe it.

We use the `MessageAdaptionFailure` signal to be able to pass an exception thrown from an adapter into the behavior stack, be rethrown and trigger the correct sequence of supervision interceptors (if there are any). 

That already worked as it should, however, if the result of handling a successfully adapted message threw an exception in the user message handling, when `rethrowTypedFailure` is true (it is for pretty much all typed actors) a `next()` on L160 would see a failed behavior and rethrow a `TypedActorFailedException` intended to trigger classic supervision, but in this case it was caught and again passed into the behavior stack triggering the wrong supervision as it is a different exception type than expected.

Reply:
Thanks for the clarification @johanandren